python/ruff-ecosystem/ruff_ecosystem/projects.py

```diff
@@ -52,6 +59,54 @@ def with_options(self: Self, **kwargs) -> Self:
     def to_ruff_args(self) -> list[str]:
         pass
 
+    @contextlib.contextmanager
+    def update_toml(self, dirpath: Path) -> None:
+        ruff_toml = dirpath / "ruff.toml"
+        pyproject_toml = dirpath / "pyproject.toml"
+        if ruff_toml.exists():
+            path = ruff_toml
+            base = []
+        else:
+            path = pyproject_toml
+            base = ["tool", "ruff"]
+
+        if self.toml_overrides and path.exists():
+            contents = path.read_text()
+            toml = tomli.loads(contents)
+
+            # Update the TOML, using `.` to descend into nested keys
+            for key, value in self.toml_overrides:
+                # Allow "preview" only overrides
+                if key.startswith("preview."):
+                    if not self.preview:
+                        continue
+                    else:
+                        key = key[len("preview.") :]
+
+                # Allow "no-preview" only overrides
+                if key.startswith("no-preview."):
+                    if self.preview:
+                        continue
+                    else:
+                        key = key[len("no-preview.") :]
```

Comment:
This seems a little confusing but something like this is needed to solve the problem.